### ISSUE TITLE
Using dataParTasksPerLocale in library code

### ISSUE THREAD

**Issue by mppf** [chapel-lang/chapel#12784]
I am working on some library code that should allow dataParTasksPerLocale to control the level of parallelism but doesn't use a forall loop.

I would use dataParTasksPerLocale directly, but it is the configuration const, rather than something based on the number of cores, so defaults to 0.

I see many distributions need to do something like this:

``` chapel
const nTasksPerLocale = if dataParTasksPerLocale>0 then dataParTasksPerLocale
                                                   else here.maxTaskPar;

writeln(nTasksPerLocale);
```

Can the standard library include a helper function to do this? Arguably getDataParTasksPerLocale could do so.

**Comment by ronawho** [comment]
I have been running into this pretty frequently, and agree it would be helpful to have a wrapper (or to have `dataParTasksPerLocale` default to `here.maxTaskPar`)

**Comment by bradcray** [comment]
I'm trying to remember (without checking) whether there was a need to have a sentinel value of 0 to mean "the user didn't specify anything" distinctly from "they happened to set it to here.maxTaskPar."  I think it may have simply been because different locales might have different numbers of tasks, so if we literally wrote:

```chapel
config const dataParTasksPerLocale = here.maxTaskPar;
```

then all locales would use a value that might only be appropriate for locale 0 (?).

**Comment by ronawho** [comment]
Oh, shoot -- yeah that's a really good point.

**Comment by mppf** [comment]
> it may have simply been because different locales might have different numbers of tasks...

Right, so couldn't we set dataParTasksPerLocale to the minimum across locales at program start up? Note I'd still consider it important for it to be const. It could be declared like this:

``` chapel
config const dataParTasksPerLocale = computeMinimumTaskParOnLocales();
```

**Comment by bradcray** [comment]
> couldn't we set dataParTasksPerLocale to the minimum across locales at program start up?

We could, the downside being that a forall loop running on a locale with more cores than that would not use them all.  It could be argued that our current iterators and distributions aren't very well set up for heterogeneous target environments today anyway (so using the extra cores may not add benefit since that locale may end up waiting for the others anyway), but I like that the current scheme sets us up for such distributions and iterators in the future even if we don't support them today.  Is it worth the pain and suffering in the meantime?  That's hard to say, but creating a library routine to get the value seems like a simple way to relieve that for now.

**Comment by mppf** [comment]
A couple notes. First, this issue is related to #17562 in that we might wish the solution here to work with serial statements.

Second, I have been using this code implementing getting the number of tasks as a workaround for this missing feature:

``` chpl
proc computeNumTasks() {
  if __primitive("task_get_serial") {
    return 1;
  }

  const tasksPerLocale = dataParTasksPerLocale;
  const ignoreRunning = dataParIgnoreRunningTasks;
  var nTasks = if tasksPerLocale > 0 then tasksPerLocale else here.maxTaskPar;
  if !ignoreRunning {
    const otherTasks = here.runningTasks() - 1; // don't include self
    nTasks = if otherTasks < nTasks then (nTasks-otherTasks):int else 1;
  }

  return nTasks;
}
```
